### ISSUE TITLE
Production Login Errors

### ISSUE THREAD

**Issue by AndyYouens** [hmrc/vat-api#680]
anyone having login problems in production this morning?

getting the following error;

{"name":"Error","status":401,"message":"Unauthorized","context":{"error":"invalid_client","error_description":"invalid client id or secret"}}

Both client ID & secret are correct.

**Comment by adamkecskes** [comment]
@AndyYouens we're having the same issue. The test environment works fine, but the prod responses this all the time. 
In the past few days we got the 'code is invalid' error response, but today we're getting 'invalid client id or secret'

**Comment by AndyYouens** [comment]
Thanks for the update @adamkecskes

**Comment by AndyYouens** [comment]
Any response HMRC?  This is a production problem but nothing heard.

As @adamkecskes says - this is not a problem in development.

**Comment by mtd-api-team** [comment]
Hi @AndyYouens 

As a reminder this channel is for technical questions related to the vat-api. And unfortunately we are unable to assist with Auth related issues. 

Judging by the error "invalid client id or secret" I'm guessing this is the oauth request to retrieve an access (bearer) token? I would encourage you to contact SDST SDSTeam@hmrc.gov.uk who will be able to help in this case. 

Regards

**Comment by AndyYouens** [comment]
@mtd-api-team 

Yes, I have contacted them & we all know how long they take to respond :-(

Not good enough for production problems!

**Comment by mtd-api-team** [comment]
Hi @AndyYouens 

We are aware of a similar issue when sending the `Authorization` header to the `/oauth/token` endpoint. This header should not be included.

Let us know if this resolves your issues.

Regards

**Comment by adamkecskes** [comment]
@mtd-api-team no, I've tried to remove the `Authorization` header from that request, but I received the same response.
How is it possible that the `Authorization` header is a problem on production, but not on test environment?

**Comment by AndyYouens** [comment]
@mtd-api-team 

I will test your theory.  Surely the error message should reflect something a bit more meaningful.

invalid client id or secret ??

This works in the sandbox!

As all the developers have stated on many occassions, why cant the sandbox operate the same as the production environment??

We should not be testing in production, but that is the only way we can test our software - all very poor!

**Comment by mtd-api-team** [comment]
@AndyYouens as mentioned, we don't look after authorisation to interact with HMRC APIs, we look after the technical elements of the VAT API.

The explanations for the error messages you are receiving can be found here: https://developer.service.hmrc.gov.uk/api-documentation/docs/authorisation/user-restricted-endpoints

As you know we try and be as helpful as we can where possible.  This time, if you email over your Client IDs to service_mtd.digital_ddcl@digital.hmrc.gov.uk, we can look into this further for you.

**Comment by AndyYouens** [comment]
any joy your end @adamkecskes ?

**Comment by adamkecskes** [comment]
It seems there is @AndyYouens ! :)
Our `/oauth/token` requests were sent with `Contenty-Type: application/x-www-form-urlencoded`.
When we modified our content to JSON, and removed the `Authorization` header, the endpoint stareted responding correctly.
It's still a mystery why their TEST endpoint accepts both JSON and urlencoded formats while their LIVE doesn't.

**Comment by AndyYouens** [comment]
We still have a problem.  Works fine in test but not in production.

Neither environments are sending the authorization header & the content-type is application/json

Waiting on a reply for SDST :-(
